restart.c

```diff
@@ -0,0 +1,400 @@
+#include "memcached.h"
+
+#include "restart.h"
+
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <string.h>
+
+typedef struct _restart_data_cb restart_data_cb;
+
+struct _restart_data_cb {
+    void *data; // user supplied opaque data.
+    struct _restart_data_cb *next; // callbacks are ordered stack
+    restart_check_cb ccb;
+    restart_save_cb scb;
+    char tag[RESTART_TAG_MAXLEN];
+};
+
+// TODO: struct to hand back to caller.
+static int mmap_fd = 0;
+static void *mmap_base = NULL;
+static size_t slabmem_limit = 0;
+char *memory_file = NULL;
+
+static restart_data_cb *cb_stack = NULL;
+
+// Allows submodules and engines to have independent check and save metadata
+// routines for the restart code.
+void restart_register(const char *tag, restart_check_cb ccb, restart_save_cb scb, void *data) {
+    restart_data_cb *cb = calloc(1, sizeof(restart_data_cb));
+
+    // Handle first time call initialization inline so we don't need separate
+    // API call.
+    if (cb_stack == NULL) {
+        cb_stack = cb;
+    } else {
+        // Ensure we fire the callbacks in registration order.
+        // Someday I'll get a queue.h overhaul.
+        restart_data_cb *finder = cb_stack;
+        while (finder->next != NULL) {
+            finder = finder->next;
+        }
+        finder->next = cb;
+    }
+
+    // TODO: Write a safe_strncpy into util.c and stop using it raw.
+    // or a local strlcpy.
+    strncpy(cb->tag, tag, RESTART_TAG_MAXLEN);
+    cb->tag[RESTART_TAG_MAXLEN-1] = '\0';
+    cb->data = data;
+    cb->ccb = *ccb;
+    cb->scb = *scb;
+}
+
+typedef struct {
+    FILE *f;
+    restart_data_cb *cb;
+    char *line;
+    bool done;
+} restart_cb_ctx;
+
+// TODO: error string from cb?
+// - look for final line with checksum
+// - checksum entire file (up until final line)
+// - seek to start
+
+static int restart_check(const char *file) {
+    // metadata is kept in a separate file.
+    size_t flen = strlen(file);
+    const char *ext = ".meta";
+    char *metafile = malloc(flen + strlen(ext));
+    memcpy(metafile, file, flen);
+    memcpy(metafile+flen, ext, strlen(ext));
+
+    FILE *f = fopen(metafile, "r");
```

Comment:
This does bother me but I gave up on fixing it right now. The init code is a bit of a rats nest.

I'd take a (really simple) suggestion, else I'll punt this until I can think about it more directly. Part of the problem here is this mode actually has more uses than being restartable, which is why it's called "memory-file" and not "restart-file". IE: for users of PMEM they would want to point item memory out into a DAX filesystem, but not necessarily care about the restart mode. Thus the restart code only ever engages when shutdown properly or if the .meta file exists.

Reply:
You're right, it does seem like it will take some effort refactor. I'm happy to look at that and submit a PR in the near future, since I did spend sometime reading this part of the code.